### PR TITLE
Use h5d_append to simplify ObservableHelper

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

I used h5d_append to remove the `value1_d` and `space1_d` state variables in ObservableHelper. Passing `hdf_archive` to `ObservableHelper::write` lets us remove `ObservableHelper::open`.

Variadic templates and initializer lists don't play nicely together, which is why `std::vector<std::string>{...} is added inside `std::make_unique(...)` and `std::vector<T>::emplace_back(...)`.   

Some writes still occur inside registerCollectables and registerObservables, so I didn't try removing `hdf_archive &` from the arguments list. 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No 

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes/No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCHamiltonians/ObservableHelper.h [hunk 1]
```diff
@@ -34,30 +34,12 @@ using value_type = QMCTraits::FullPrecRealType;
...
-  std::vector<hsize_t> mydims;
-  ///maximum dimensions
-  std::vector<hsize_t> maxdims;
-  ///current dimensions while extending data
-  std::vector<hsize_t> curdims;
-  ///offsets
-  std::vector<hsize_t> offsets;
-  ///name of this observable
-  std::string group_name;
-
```
Comment (ye-luo) [inline src/QMCHamiltonians/ObservableHelper.h:56]: Getting rid of the above is a clear win.

File: src/QMCHamiltonians/ObservableHelper.h [hunk 2]
```diff
@@ -109,21 +87,19 @@ class ObservableHelper
...
 private:
-  ///true: hdf5 handlers are valid via open, false: otherwise
-  bool isopened = false;
-
-  ///closes remaining hdf5 handlers in destructor if isopened = true
-  void close();
+  /// "file pointer" for h5d_append
+  hsize_t current = 0;
+  ///name of this observable
+  std::vector<std::string> group_name;
```
Comment (ye-luo) [inline src/QMCHamiltonians/ObservableHelper.h:98]: Could you say a bit more? "vector" seems surprising. Also the variable name is not plural.

Comment (quantumsteve) [reply src/QMCHamiltonians/ObservableHelper.h:98]: `hdf_archive` isn't in the same top-level group when `ObservableHelper::open(...)` and `ObservableHelper::write(...)` are called. We need to store a path containing multiple groups. I initially tried writing a longer string where each group was separated by `/` but HDF5 wouldn't create multiple groups at once. I found it easier to iterate through a vector of strings than than do the string manipulation to generate these substrings.

Essentially, I'm iterating over a a HDF5 path similar to how you might do on the fileystem with [std::filesystem::path]( 
https://en.cppreference.com/w/cpp/filesystem/path/begin)

File: src/QMCHamiltonians/DensityEstimator.cpp [hunk 3]
```diff
@@ -13,7 +13,7 @@
 //                    Mark A. Berrill, berrillma@ornl.gov, Oak Ridge National Laboratory
 //
 // File created by: Bryan Clark, bclark@Princeton.edu, Princeton University
-//////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////.///////////
```
Comment (markdewing) [inline src/QMCHamiltonians/DensityEstimator.cpp:16]: Added character (accidental, I assume)

File: src/QMCHamiltonians/ObservableHelper.h (hunk 2 above)
Comment (ye-luo) [reply src/QMCHamiltonians/ObservableHelper.h:98]: Yes. That is useful info. Could you expand the comment saying the groups that a patch consists of is stored here.

File: src/QMCHamiltonians/DensityEstimator.cpp (hunk 3 above)
Comment (quantumsteve) [reply src/QMCHamiltonians/DensityEstimator.cpp]: :facepalm:

**Review by ye-luo — APPROVED** [review]
I couldn't manage to understand every change here but the code is neat.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**